### PR TITLE
add support for specifying secondary indexes with to_sql

### PR DESCRIPTION
**Pull request by trbs**
Support for creating secondary indexes in `to_sql` via the parameter `indexes`.

The code also tries to avoid creating duplicate indexes when `keys` is specified in `SQLTable`.
(Once the PR for passing the `keys` this would be useful)

This PR introduces a new method in `SQLTable` called `_is_column_indexed` so that subclasses can easily override or change the logic.

Interesting future work will be to also support creating compound indexes.

### CODE REVIEW

File: pandas/io/sql.py [hunk 1]
```diff
@@ -516,7 +516,8 @@ def read_sql(sql, con, index_col=None, coerce_float=True, params=None,
 
 
 def to_sql(frame, name, con, flavor='sqlite', schema=None, if_exists='fail',
-           index=True, index_label=None, chunksize=None, dtype=None):
+           index=True, index_label=None, chunksize=None, dtype=None,
+           indexes=None):
```
Comment (jreback) [inline pandas/io/sql.py]: would need to be added to the doc-string with a versionadded directive

File: pandas/io/sql.py [hunk 2]
```diff
@@ -849,18 +851,33 @@ def _index_name(self, index, index_label):
         else:
             return None
 
+    def _is_column_indexed(self, label):
+        if self.indexes is not None and label in self.indexes:
```
Comment (jreback) [inline pandas/io/sql.py]: put a comment explaining this logic

File: pandas/io/tests/test_sql.py [hunk 3]
```diff
@@ -966,6 +975,66 @@ def test_to_sql_read_sql_with_database_uri(self):
         tm.assert_frame_equal(test_frame1, test_frame3)
         tm.assert_frame_equal(test_frame1, test_frame4)
 
+    def test_to_sql_column_indexes(self):
+        temp_frame = DataFrame({'col1': range(4), 'col2': range(4)})
```
Comment (jreback) [inline pandas/io/tests/test_sql.py]: add the issue number as a comment

**Comment by jreback** [comment]
does this have an open issue for this? (IIRC there is)

**Comment by trbs** [comment]
I didn't find this exact issue.

https://github.com/pydata/pandas/issues/7984 and https://github.com/pydata/pandas/issues/9084 are similar in that they also deal with to_sql / sql indexes but not specifically specifying which columns should be indexes.

I am interested in maybe picking up https://github.com/pydata/pandas/issues/7984 as well after this PR. Together with a PR for specifying multi-column / compound indexes.

File: pandas/io/sql.py (hunk 1 above)
Comment (trbs) [inline pandas/io/sql.py]: Can this be included in v0.18.1 ? or will it be for v0.19.0.

I will also update the corresponding whatsnew/.txt file.

**Comment by jreback** [comment]
hmm @trbs this PR looks like it should close those 2 issues, or is there something that is indicated in those that is not in your PR?

**Comment by trbs** [comment]
#7984 is about unique indexes... this PR would only add an index to the column not unique. I add that to this PR.

#9084 is basically about adding more parameters to get_schema(...) to that it includes the index=.... and index_label=... parameters that to_sql has. (so its about the dataframe index not database indices)

I could do the patch for #9084 as will since it looks kinda trivial. But it's probably better to do that in a separate PR ?

**Comment by jreback** [comment]
@jorisvandenbossche

**Comment by jreback** [comment]
status? @jorisvandenbossche

**Comment by jreback** [comment]
can you rebase / update?

File: pandas/core/generic.py [hunk 4]
```diff
@@ -1158,7 +1158,8 @@ def to_msgpack(self, path_or_buf=None, encoding='utf-8', **kwargs):
                                   **kwargs)
 
     def to_sql(self, name, con, flavor=None, schema=None, if_exists='fail',
-               index=True, index_label=None, chunksize=None, dtype=None):
+               index=True, index_label=None, chunksize=None, dtype=None,
+               indexes=None):
         """
         Write records stored in a DataFrame to a SQL database.
```
Comment (jreback) [inline pandas/core/generic.py]: are the params documented here?

File: pandas/io/sql.py [hunk 5]
```diff
@@ -445,6 +446,10 @@ def to_sql(frame, name, con, flavor=None, schema=None, if_exists='fail',
         Column label for index column(s). If None is given (default) and
         `index` is True, then the index names are used.
         A sequence should be given if the DataFrame uses MultiIndex.
+    indexes : list of column name(s). Columns names in this list will have
+        an indexes created for them in the database.
+
+        .. versionadded:: 0.18.2
```
Comment (jreback) [inline pandas/io/sql.py]: 0.20.0

**Comment by jreback** [comment]
@jorisvandenbossche status?

**Comment by trbs** [comment]
btw happy to rebase again if Pandas wants to merge this PR.

**Comment by jreback** [comment]
closing as stale

**Pull request closed without merging**